File: mp3_player.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// -----------------------------------------------------------------------------
// Cross-Platform Definitions for Audio Playback Command
// -----------------------------------------------------------------------------

#ifdef _WIN32
    // Windows: Use 'start' to open the file with the default media player
    #define PLAY_COMMAND_FORMAT "start \"\" \"%s\""
    #define PLATFORM_NOTE "Using Windows 'start' command."
#elif defined(__APPLE__)
    // macOS: Use 'afplay' (built-in command line audio player)
    #define PLAY_COMMAND_FORMAT "afplay \"%s\""
    #define PLATFORM_NOTE "Using macOS 'afplay' command."
#else
    // Linux/Other POSIX: Use 'mpg123' (a common command-line MP3 player)
    // NOTE: This requires the 'mpg123' utility to be installed (e.g., 'sudo apt install mpg123').
    #define PLAY_COMMAND_FORMAT "mpg123 \"%s\""
    #define PLATFORM_NOTE "Using Linux 'mpg123' command (requires installation)."
#endif

// Define the name of the MP3 file the program looks for
#define MP3_FILE_NAME "audio.mp3"
#define MAX_COMMAND_LENGTH 512

/**
 * Main function to execute the playback command.
 */
int main() {
    char command[MAX_COMMAND_LENGTH];
    int result;

    printf("======================================\n");
    printf("      C Command-Line MP3 Player       \n");
    printf("======================================\n");

    printf("Platform Detected: %s\n", PLATFORM_NOTE);
    printf("Attempting to play: %s\n", MP3_FILE_NAME);

    // 1. Construct the system command string
    // snprintf is used for safe string formatting to prevent buffer overflows
    if (snprintf(command, MAX_COMMAND_LENGTH, PLAY_COMMAND_FORMAT, MP3_FILE_NAME) >= MAX_COMMAND_LENGTH) {
        fprintf(stderr, "Error: Command string is too long.\n");
        return 1;
    }

    printf("Executing command: %s\n", command);

    // 2. Execute the system command
    // The system() function executes the command and waits for it to complete.
    // If the external player opens in a new window, this program will continue immediately.
    result = system(command);

    // 3. Check the result of the system call
    if (result == -1) {
        fprintf(stderr, "\nERROR: Failed to execute system command (system function call failed).\n");
        return 1;
    } else if (result != 0) {
        // A non-zero result usually indicates an error with the command execution
        fprintf(stderr, "\nERROR: Playback command failed with exit code %d.\n", result);
        fprintf(stderr, "Possible causes:\n");
        fprintf(stderr, "  1. The file '%s' is missing or misspelled.\n", MP3_FILE_NAME);
        #ifdef __linux__
            fprintf(stderr, "  2. The 'mpg123' player is not installed.\n");
            fprintf(stderr, "     (Try: 'sudo apt install mpg123' or 'sudo yum install mpg123')\n");
        #else
            fprintf(stderr, "  2. The necessary system player is unavailable.\n");
        #endif
        return 1;
    }

    printf("\nPlayback command executed. Check your desktop or console for the audio player.\n");
    printf("--------------------------------------\n");

    return 0;
}

File: mp3_player.py
import subprocess
import sys
import os

# --- Configuration ---
MP3_FILE_NAME = "audio.mp3"
DEFAULT_MESSAGE = "Check your desktop or console for the audio player."

def play_audio(filepath):
    """
    Plays the specified audio file using a platform-appropriate command.
    """
    platform = sys.platform

    if not os.path.exists(filepath):
        print(f"\n[ERROR] File not found: '{filepath}'")
        print("Please ensure your MP3 file is in the same directory and named exactly 'audio.mp3'.")
        return

    command = None
    player_note = ""

    # 1. Determine the command based on the operating system
    if platform.startswith('win'):
        # Windows: Use 'start' to open the file with the default media player
        command = ['start', '', filepath]
        player_note = "Using Windows 'start' command."
    elif platform.startswith('darwin'):
        # macOS: Use 'afplay' (built-in command line audio player)
        command = ['afplay', filepath]
        player_note = "Using macOS 'afplay' command."
    elif platform.startswith('linux'):
        # Linux: Try 'mpg123' first, then 'aplay' or 'xdg-open'
        # mpg123 is excellent for MP3s but often requires installation.
        command = ['mpg123', filepath]
        player_note = "Using Linux 'mpg123' command (may require installation)."
    else:
        print(f"[ERROR] Unsupported platform: {platform}")
        return

    print("=======================================")
    print("      Python Command-Line Player       ")
    print("=======================================")
    print(f"Platform Detected: {player_note}")
    print(f"Attempting to play: {filepath}")
    print(f"Executing: {' '.join(command)}")

    # 2. Execute the command
    try:
        # Use subprocess.Popen for non-blocking execution (especially on Windows/macOS)
        # On Linux, mpg123 will block unless we run it in the background, but this is simpler.
        # The 'start' command on Windows and 'afplay' on macOS generally allow the script to continue.
        subprocess.Popen(command, shell=(platform.startswith('win')))
        print(f"\nSuccess! {DEFAULT_MESSAGE}")

    except FileNotFoundError:
        # This occurs if the command itself (e.g., 'mpg123' or 'afplay') is not found
        print("\n[ERROR] Playback utility not found.")
        if platform.startswith('linux'):
            print("If you are on Linux, ensure 'mpg123' is installed (e.g., 'sudo apt install mpg123').")
        print("Please verify your system's required audio playback tools are accessible.")
    except Exception as e:
        print(f"\n[CRITICAL ERROR] Failed to execute playback command: {e}")

if __name__ == "__main__":
    # Get the directory of the script and construct the full path to the MP3 file
    script_dir = os.path.dirname(os.path.abspath(sys.argv[0]))
    audio_path = os.path.join(script_dir, MP3_FILE_NAME)
    
    play_audio(audio_path)

